src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,300 @@
+// Other functions of this class should be labelled as private.
+template<typename MatType, typename DataOptionsType>
 bool Load(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          DatasetMapper<PolicyType>& info,
-          const bool fatal,
-          const bool transpose)
+          MatType& matrix,
+          DataOptionsType& opts,
+          std::enable_if_t<IsArma<MatType>::value || IsSparseMat<MatType>::value>*,
+          std::enable_if_t<!std::is_same_v<DataOptionsType, bool>>*)
 {
-  // Get the extension and load as necessary.
   Timer::Start("loading_data");
 
-  // Get the extension.
-  std::string extension = Extension(filename);
-
-  // Catch nonexistent files by opening the stream ourselves.
   std::fstream stream;
-  stream.open(filename.c_str(), std::fstream::in);
-
-  if (!stream.is_open())
+  bool success = OpenFile(filename, opts, true, stream);
+  if (!success)
   {
     Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
-
     return false;
   }
 
-  if (extension == "csv" || extension == "tsv" || extension == "txt")
+  success = DetectFileType<MatType>(filename, opts, true, &stream);
+  if (!success)
   {
-    Log::Info << "Loading '" << filename << "' as CSV dataset.  " << std::flush;
-    try
-    {
-      LoadCSV loader(filename);
-      loader.LoadCategoricalCSV(matrix, info, transpose);
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
+    Timer::Stop("loading_data");
+    return false;
+  }
 
-      return false;
-    }
+  if constexpr (IsArma<MatType>::value || IsSparseMat<MatType>::value)
+  {
+    // compile time type detction
+    // convert this to a move, and move back at the end.
+    // add a move constructor that does this.
+    TextOptions txtOpts(std::move(opts));
+    success = LoadMatrix(filename, matrix, stream, txtOpts);
+    opts = std::move(txtOpts);
   }
-  else if (extension == "arff")
+  else 
   {
-    Log::Info << "Loading '" << filename << "' as ARFF dataset.  "
-        << std::flush;
-    try
-    {
-      LoadARFF(filename, matrix, info);
-
-      // We transpose by default.  So, un-transpose if necessary...
-      if (!transpose)
-      {
-        return inplace_transpose(matrix, fatal);
-      }
-    }
-    catch (std::exception& e)
-    {
-      Timer::Stop("loading_data");
-      if (fatal)
-        Log::Fatal << e.what() << std::endl;
-      else
-        Log::Warn << e.what() << std::endl;
-
-      return false;
-    }
+    throw std::runtime_error("DataOptionType is unknown!."
+        "please use a known type or provide specific overloads");
   }
-  else
+
+  if (!success)
   {
-    // The type is unknown.
+    Log::Info << std::endl;
```

Comment:
```suggestion
```

Seems like this line wouldn't do anything other than print `[INFO ]` with no text?